File: lld/week-02-03-patterns/proxy/go/proxy.go
// Package proxy demonstrates the Proxy pattern: BankAccountProxy controls
// access to a RealBankAccount, enforcing owner authorization before any
// withdrawal or deposit reaches the real object.
package proxy

import (
	"errors"
	"fmt"
)

var ErrUnauthorized = errors.New("unauthorized: requester is not the account owner")
var ErrInsufficientFunds = errors.New("insufficient funds")

// Account is the common interface shared by the real subject and its proxy,
// so callers can use either interchangeably.
type Account interface {
	Deposit(requester string, amount float64) error
	Withdraw(requester string, amount float64) error
	Balance() float64
}

// RealBankAccount is the expensive/sensitive real subject. It performs no
// authorization checks of its own — that's the proxy's job.
type RealBankAccount struct {
	Owner   string
	balance float64
}

func NewRealBankAccount(owner string, initialBalance float64) *RealBankAccount {
	return &RealBankAccount{Owner: owner, balance: initialBalance}
}

func (a *RealBankAccount) Deposit(_ string, amount float64) error {
	a.balance += amount
	return nil
}

func (a *RealBankAccount) Withdraw(_ string, amount float64) error {
	if amount > a.balance {
		return ErrInsufficientFunds
	}
	a.balance -= amount
	return nil
}

func (a *RealBankAccount) Balance() float64 { return a.balance }

// BankAccountProxy is a protection proxy: it enforces that only the account
// owner may deposit or withdraw, and logs every access attempt, without the
// real account needing to know about authorization at all.
type BankAccountProxy struct {
	real      *RealBankAccount
	AccessLog []string
}

func NewBankAccountProxy(real *RealBankAccount) *BankAccountProxy {
	return &BankAccountProxy{real: real}
}

func (p *BankAccountProxy) Deposit(requester string, amount float64) error {
	p.AccessLog = append(p.AccessLog, fmt.Sprintf("deposit attempt by %s", requester))
	if requester != p.real.Owner {
		return ErrUnauthorized
	}
	return p.real.Deposit(requester, amount)
}

func (p *BankAccountProxy) Withdraw(requester string, amount float64) error {
	p.AccessLog = append(p.AccessLog, fmt.Sprintf("withdraw attempt by %s", requester))
	if requester != p.real.Owner {
		return ErrUnauthorized
	}
	return p.real.Withdraw(requester, amount)
}

func (p *BankAccountProxy) Balance() float64 { return p.real.Balance() }

File: lld/week-02-03-patterns/proxy/python/proxy.py
"""Proxy pattern — BankAccountProxy controls access to a RealBankAccount.

See ../README.md for the design writeup.
"""
from __future__ import annotations

from typing import List, Protocol


class UnauthorizedError(Exception):
    pass


class InsufficientFundsError(Exception):
    pass


class Account(Protocol):
    def deposit(self, requester: str, amount: float) -> None: ...
    def withdraw(self, requester: str, amount: float) -> None: ...
    def balance(self) -> float: ...


class RealBankAccount:
    """The expensive/sensitive real subject. Performs no authorization
    checks of its own — that's the proxy's job."""

    def __init__(self, owner: str, initial_balance: float):
        self.owner = owner
        self._balance = initial_balance

    def deposit(self, requester: str, amount: float) -> None:
        self._balance += amount

    def withdraw(self, requester: str, amount: float) -> None:
        if amount > self._balance:
            raise InsufficientFundsError("insufficient funds")
        self._balance -= amount

    def balance(self) -> float:
        return self._balance


class BankAccountProxy:
    """Protection proxy: enforces that only the account owner may deposit or
    withdraw, and logs every access attempt."""

    def __init__(self, real: RealBankAccount):
        self._real = real
        self.access_log: List[str] = []

    def deposit(self, requester: str, amount: float) -> None:
        self.access_log.append(f"deposit attempt by {requester}")
        if requester != self._real.owner:
            raise UnauthorizedError("requester is not the account owner")
        self._real.deposit(requester, amount)

    def withdraw(self, requester: str, amount: float) -> None:
        self.access_log.append(f"withdraw attempt by {requester}")
        if requester != self._real.owner:
            raise UnauthorizedError("requester is not the account owner")
        self._real.withdraw(requester, amount)

    def balance(self) -> float:
        return self._real.balance()


def _demo() -> None:
    real = RealBankAccount("alice", 100)
    acc = BankAccountProxy(real)

    acc.withdraw("alice", 40)
    print(f"Balance after owner withdraw: {acc.balance()}")

    try:
        acc.withdraw("mallory", 10)
    except UnauthorizedError as e:
        print(f"Blocked: {e}")


if __name__ == "__main__":
    _demo()
